From c2df403bd1735e366fb0be558671adbdf9c4a4bd Mon Sep 17 00:00:00 2001
From: jizhenlo <zhenlong.z.ji@intel.com>
Date: Thu, 9 Nov 2023 16:59:53 +0800
Subject: [PATCH] Modify systemupdatersample to support OTA from U disk

Customers can do OTA from U disk through this app.

Tracked-On: OAM-113303
Signed-off-by: jizhenlo <zhenlong.z.ji@intel.com>
---
 updater_sample/Android.bp                     |   2 +
 updater_sample/AndroidManifest.xml            |  10 +-
 updater_sample/res/layout/activity_main.xml   |  98 ++++++---------
 updater_sample/res/values/strings.xml         |   5 +-
 .../systemupdatersample/UpdateConfig.java     |   5 +
 .../systemupdatersample/UpdateManager.java    |  16 ++-
 .../systemupdatersample/UpdaterState.java     |  13 +-
 .../services/PrepareUpdateService.java        |  10 ++
 .../systemupdatersample/ui/MainActivity.java  | 118 +++++++++---------
 .../util/UpdateConfigs.java                   |  71 +++++++----
 10 files changed, 199 insertions(+), 149 deletions(-)

diff --git a/updater_sample/Android.bp b/updater_sample/Android.bp
index 9222d063..b741f9c7 100644
--- a/updater_sample/Android.bp
+++ b/updater_sample/Android.bp
@@ -38,4 +38,6 @@ android_app {
     },
 
     resource_dirs: ["res"],
+
+	certificate: "platform",
 }
diff --git a/updater_sample/AndroidManifest.xml b/updater_sample/AndroidManifest.xml
index 981cd8eb..41b5c9da 100644
--- a/updater_sample/AndroidManifest.xml
+++ b/updater_sample/AndroidManifest.xml
@@ -17,10 +17,11 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
           package="com.example.android.systemupdatersample">
 
-    <uses-sdk android:minSdkVersion="27" android:targetSdkVersion="27" />
-
-    <uses-permission android:name="android.permission.ACCESS_CACHE_FILESYSTEM" />
     <uses-permission android:name="android.permission.INTERNET"/>
+    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
+    <uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE"/>
+	<uses-permission android:name="android.permission.REBOOT" />
+	<uses-permission android:name="android.permission.DEVICE_POWER" />
 
     <application
         android:icon="@mipmap/ic_launcher"
@@ -28,7 +29,8 @@
         android:roundIcon="@mipmap/ic_launcher_round">
         <activity
             android:name=".ui.MainActivity"
-            android:label="@string/app_name" >
+			android:label="@string/app_name"
+			android:exported="true" >
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
                 <category android:name="android.intent.category.LAUNCHER" />
diff --git a/updater_sample/res/layout/activity_main.xml b/updater_sample/res/layout/activity_main.xml
index b560827d..aa87ec2e 100644
--- a/updater_sample/res/layout/activity_main.xml
+++ b/updater_sample/res/layout/activity_main.xml
@@ -39,7 +39,7 @@
                 android:id="@+id/textViewBuildtitle"
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
-                android:text="Current Build:" />
+                android:text="当前系统:" />
 
             <TextView
                 android:id="@+id/textViewBuild"
@@ -51,27 +51,13 @@
                 android:layout_width="match_parent"
                 android:layout_height="40dp" />
 
-            <TextView
-                android:id="@+id/textView4"
-                android:layout_width="match_parent"
-                android:layout_height="wrap_content"
-                android:text="Apply an update" />
-
-            <TextView
-                android:id="@+id/textViewConfigsDirHint"
-                android:layout_width="match_parent"
-                android:layout_height="wrap_content"
-                android:layout_marginTop="4dp"
-                android:text="Config files located in NULL"
-                android:textColor="#777"
-                android:textSize="10sp"
-                android:textStyle="italic" />
-
-            <Spinner
-                android:id="@+id/spinnerConfigs"
-                android:layout_width="match_parent"
-                android:layout_height="wrap_content"
-                android:layout_marginTop="8dp" />
+			<Button
+				android:id="@+id/buttonReload"
+				android:layout_width="wrap_content"
+				android:layout_height="wrap_content"
+				android:layout_marginTop="4dp"
+				android:onClick="onReloadClick"
+				android:text="刷新" />
 
             <LinearLayout
                 android:layout_width="match_parent"
@@ -79,31 +65,28 @@
                 android:layout_marginTop="8dp"
                 android:orientation="horizontal">
 
-                <Button
-                    android:id="@+id/buttonReload"
-                    android:layout_width="0dp"
-                    android:layout_height="wrap_content"
-                    android:layout_weight="1"
-                    android:onClick="onReloadClick"
-                    android:text="Reload" />
-
-                <Button
-                    android:id="@+id/buttonViewConfig"
-                    android:layout_width="0dp"
-                    android:layout_height="wrap_content"
-                    android:layout_weight="1"
-                    android:onClick="onViewConfigClick"
-                    android:text="View config" />
-
-                <Button
-                    android:id="@+id/buttonApplyConfig"
-                    android:layout_width="0dp"
-                    android:layout_height="wrap_content"
-                    android:layout_weight="1"
-                    android:onClick="onApplyConfigClick"
-                    android:text="Apply" />
+				<TextView
+					android:id="@+id/textView4"
+					android:layout_width="wrap_content"
+					android:layout_height="wrap_content"
+					android:text="升级包:" />
+
+				<TextView
+					android:id="@+id/textViewConfigsDirHint"
+					android:layout_width="wrap_content"
+					android:layout_height="wrap_content"
+					android:layout_marginLeft="8dp"
+					android:text="@string/package_info" />
             </LinearLayout>
 
+			<Button
+				android:id="@+id/buttonApplyConfig"
+				android:layout_width="match_parent"
+				android:layout_height="wrap_content"
+				android:layout_marginTop="40dp"
+				android:onClick="onApplyConfigClick"
+				android:text="安装更新" />
+
             <LinearLayout
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
@@ -161,23 +144,23 @@
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
                     android:layout_marginLeft="8dp"
-                    android:text="@string/unknown" />
+                    android:text="@string/none" />
             </LinearLayout>
 
 
             <ProgressBar
                 android:id="@+id/progressBar"
                 style="?android:attr/progressBarStyleHorizontal"
-                android:layout_marginTop="8dp"
+                android:layout_marginTop="40dp"
                 android:min="0"
-                android:max="100"
+				android:max="100"
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content" />
 
             <LinearLayout
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
-                android:layout_marginTop="12dp"
+                android:layout_marginTop="40dp"
                 android:orientation="horizontal">
 
                 <Button
@@ -186,7 +169,7 @@
                     android:layout_height="wrap_content"
                     android:layout_weight="1"
                     android:onClick="onStopClick"
-                    android:text="Stop" />
+                    android:text="停止更新" />
 
                 <Button
                     android:id="@+id/buttonReset"
@@ -194,7 +177,7 @@
                     android:layout_height="wrap_content"
                     android:layout_weight="1"
                     android:onClick="onResetClick"
-                    android:text="Reset" />
+                    android:text="重置更新" />
             </LinearLayout>
 
             <LinearLayout
@@ -209,7 +192,7 @@
                     android:layout_height="wrap_content"
                     android:layout_weight="1"
                     android:onClick="onSuspendClick"
-                    android:text="Suspend" />
+                    android:text="暂停更新" />
 
                 <Button
                     android:id="@+id/buttonResume"
@@ -217,7 +200,7 @@
                     android:layout_height="wrap_content"
                     android:layout_weight="1"
                     android:onClick="onResumeClick"
-                    android:text="Resume" />
+                    android:text="恢复更新" />
             </LinearLayout>
 
             <TextView
@@ -225,17 +208,14 @@
                 android:layout_width="wrap_content"
                 android:layout_height="wrap_content"
                 android:layout_marginTop="14dp"
-                android:textColor="#777"
-                android:textSize="10sp"
-                android:textStyle="italic"
                 android:text="@string/finish_update_info" />
 
             <Button
-                android:id="@+id/buttonSwitchSlot"
+                android:id="@+id/buttonRebootDevice"
                 android:layout_width="wrap_content"
                 android:layout_height="match_parent"
-                android:onClick="onSwitchSlotClick"
-                android:text="@string/switch_slot" />
+                android:onClick="onRebootDeviceClick"
+                android:text="@string/reboot_device" />
 
         </LinearLayout>
 
diff --git a/updater_sample/res/values/strings.xml b/updater_sample/res/values/strings.xml
index db4a5dc6..c2866b94 100644
--- a/updater_sample/res/values/strings.xml
+++ b/updater_sample/res/values/strings.xml
@@ -17,7 +17,10 @@
     <string name="app_name">SystemUpdaterSample</string>
     <string name="action_reload">Reload</string>
     <string name="unknown">Unknown</string>
+    <string name="none">None</string>
     <string name="close">CLOSE</string>
     <string name="switch_slot">Switch Slot</string>
-    <string name="finish_update_info">To finish the update press the button below</string>
+    <string name="reboot_device">重启系统</string>
+    <string name="finish_update_info">更新完成，请重启以切换到新系统</string>
+    <string name="package_info">未发现升级包，请插入U盘并点击 Refresh 按钮</string>
 </resources>
diff --git a/updater_sample/src/com/example/android/systemupdatersample/UpdateConfig.java b/updater_sample/src/com/example/android/systemupdatersample/UpdateConfig.java
index 61872a63..359912dd 100644
--- a/updater_sample/src/com/example/android/systemupdatersample/UpdateConfig.java
+++ b/updater_sample/src/com/example/android/systemupdatersample/UpdateConfig.java
@@ -130,6 +130,11 @@ public class UpdateConfig implements Parcelable {
         this.mName = name;
         this.mUrl = url;
         this.mAbInstallType = installType;
+        this.mAbConfig = new AbConfig(
+                true,
+                false,
+                null,
+                null);
     }
 
     public String getName() {
diff --git a/updater_sample/src/com/example/android/systemupdatersample/UpdateManager.java b/updater_sample/src/com/example/android/systemupdatersample/UpdateManager.java
index c02e6084..6ac0fe95 100644
--- a/updater_sample/src/com/example/android/systemupdatersample/UpdateManager.java
+++ b/updater_sample/src/com/example/android/systemupdatersample/UpdateManager.java
@@ -207,7 +207,11 @@ public class UpdateManager {
     public synchronized void suspend() throws UpdaterState.InvalidTransitionException {
         Log.d(TAG, "suspend invoked");
         setUpdaterState(UpdaterState.PAUSED);
-        mUpdateEngine.cancel();
+		try {
+			mUpdateEngine.cancel();
+		} catch (Exception e) {
+            Log.e(TAG, "Failed to cancel update", e);
+		}
     }
 
     /**
@@ -263,7 +267,11 @@ public class UpdateManager {
     public synchronized void cancelRunningUpdate() throws UpdaterState.InvalidTransitionException {
         Log.d(TAG, "cancelRunningUpdate invoked");
         setUpdaterState(UpdaterState.IDLE);
-        mUpdateEngine.cancel();
+		try {
+			mUpdateEngine.cancel();
+		} catch (Exception e) {
+            Log.e(TAG, "Failed to cancel update", e);
+		}
     }
 
     /**
@@ -301,7 +309,7 @@ public class UpdateManager {
         PrepareUpdateService.startService(context, config, mHandler, (code, payloadSpec) -> {
             if (code != PrepareUpdateService.RESULT_CODE_SUCCESS) {
                 Log.e(TAG, "PrepareUpdateService failed, result code is " + code);
-                setUpdaterStateSilent(UpdaterState.ERROR);
+                setUpdaterStateSilent(code);
                 return;
             }
             updateEngineApplyPayload(UpdateData.builder()
@@ -513,6 +521,8 @@ public class UpdateManager {
             setUpdaterStateSilent(isManualSwitchSlotRequired()
                     ? UpdaterState.SLOT_SWITCH_REQUIRED
                     : UpdaterState.REBOOT_REQUIRED);
+        } else if (errorCode == UpdaterState.TIMESTAMP_ERROR) {
+            setUpdaterStateSilent(UpdaterState.TIMESTAMP_ERROR);
         } else if (errorCode != UpdateEngineErrorCodes.USER_CANCELLED) {
             setUpdaterStateSilent(UpdaterState.ERROR);
         }
diff --git a/updater_sample/src/com/example/android/systemupdatersample/UpdaterState.java b/updater_sample/src/com/example/android/systemupdatersample/UpdaterState.java
index 4eb0b68c..8c5385db 100644
--- a/updater_sample/src/com/example/android/systemupdatersample/UpdaterState.java
+++ b/updater_sample/src/com/example/android/systemupdatersample/UpdaterState.java
@@ -34,6 +34,9 @@ public class UpdaterState {
     public static final int PAUSED = 3;
     public static final int SLOT_SWITCH_REQUIRED = 4;
     public static final int REBOOT_REQUIRED = 5;
+    public static final int FILE_ERROR = 6;
+    public static final int ZIP_ERROR = 7;
+    public static final int TIMESTAMP_ERROR = 51;
 
     private static final SparseArray<String> STATE_MAP = new SparseArray<>();
 
@@ -44,6 +47,9 @@ public class UpdaterState {
         STATE_MAP.put(3, "PAUSED");
         STATE_MAP.put(4, "SLOT_SWITCH_REQUIRED");
         STATE_MAP.put(5, "REBOOT_REQUIRED");
+        STATE_MAP.put(6, "未发现有效升级包");
+        STATE_MAP.put(7, "升级包损坏，请重新拷贝升级包");
+        STATE_MAP.put(51, "升级包版本老于当前系统版本");
     }
 
     /**
@@ -52,13 +58,16 @@ public class UpdaterState {
      */
     private static final ImmutableMap<Integer, ImmutableSet<Integer>> TRANSITIONS =
             ImmutableMap.<Integer, ImmutableSet<Integer>>builder()
-                    .put(IDLE, ImmutableSet.of(IDLE, ERROR, RUNNING))
+                    .put(IDLE, ImmutableSet.of(IDLE, ERROR, ZIP_ERROR, TIMESTAMP_ERROR, RUNNING))
                     .put(ERROR, ImmutableSet.of(IDLE))
                     .put(RUNNING, ImmutableSet.of(
-                            IDLE, ERROR, PAUSED, REBOOT_REQUIRED, SLOT_SWITCH_REQUIRED))
+                            IDLE, ERROR, PAUSED, REBOOT_REQUIRED, SLOT_SWITCH_REQUIRED, FILE_ERROR, ZIP_ERROR, TIMESTAMP_ERROR))
                     .put(PAUSED, ImmutableSet.of(ERROR, RUNNING, IDLE))
                     .put(SLOT_SWITCH_REQUIRED, ImmutableSet.of(ERROR, REBOOT_REQUIRED, IDLE))
                     .put(REBOOT_REQUIRED, ImmutableSet.of(IDLE))
+                    .put(FILE_ERROR, ImmutableSet.of(IDLE))
+                    .put(ZIP_ERROR, ImmutableSet.of(IDLE))
+                    .put(TIMESTAMP_ERROR, ImmutableSet.of(IDLE))
                     .build();
 
     private AtomicInteger mState;
diff --git a/updater_sample/src/com/example/android/systemupdatersample/services/PrepareUpdateService.java b/updater_sample/src/com/example/android/systemupdatersample/services/PrepareUpdateService.java
index 29eb13da..5b977f7f 100644
--- a/updater_sample/src/com/example/android/systemupdatersample/services/PrepareUpdateService.java
+++ b/updater_sample/src/com/example/android/systemupdatersample/services/PrepareUpdateService.java
@@ -40,12 +40,14 @@ import com.example.android.systemupdatersample.util.UpdateConfigs;
 import com.google.common.collect.ImmutableSet;
 
 import java.io.File;
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Arrays;
 import java.util.Optional;
+import java.util.zip.ZipException;
 
 /**
  * This IntentService will download/extract the necessary files from the package zip
@@ -62,6 +64,8 @@ public class PrepareUpdateService extends IntentService {
      */
     public static final int RESULT_CODE_SUCCESS = 0;
     public static final int RESULT_CODE_ERROR = 1;
+    public static final int RESULT_CODE_FILE_ERROR = 6;
+    public static final int RESULT_CODE_ZIP_ERROR = 7;
 
     /**
      * Extra params that will be sent to IntentService.
@@ -131,6 +135,12 @@ public class PrepareUpdateService extends IntentService {
         try {
             PayloadSpec spec = execute(config);
             resultReceiver.send(RESULT_CODE_SUCCESS, CallbackResultReceiver.createBundle(spec));
+        } catch (FileNotFoundException e) {
+            Log.e(TAG, "Failed to prepare streaming update", e);
+            resultReceiver.send(RESULT_CODE_FILE_ERROR, null);
+        } catch (ZipException e) {
+            Log.e(TAG, "Failed to prepare streaming update", e);
+            resultReceiver.send(RESULT_CODE_ZIP_ERROR, null);
         } catch (Exception e) {
             Log.e(TAG, "Failed to prepare streaming update", e);
             resultReceiver.send(RESULT_CODE_ERROR, null);
diff --git a/updater_sample/src/com/example/android/systemupdatersample/ui/MainActivity.java b/updater_sample/src/com/example/android/systemupdatersample/ui/MainActivity.java
index 6d1e4c35..b5abd530 100644
--- a/updater_sample/src/com/example/android/systemupdatersample/ui/MainActivity.java
+++ b/updater_sample/src/com/example/android/systemupdatersample/ui/MainActivity.java
@@ -18,17 +18,18 @@ package com.example.android.systemupdatersample.ui;
 
 import android.app.Activity;
 import android.app.AlertDialog;
+import android.content.Context;
 import android.graphics.Color;
 import android.os.Build;
 import android.os.Bundle;
 import android.os.Handler;
+import android.os.PowerManager;
 import android.os.UpdateEngine;
 import android.util.Log;
 import android.view.View;
 import android.widget.ArrayAdapter;
 import android.widget.Button;
 import android.widget.ProgressBar;
-import android.widget.Spinner;
 import android.widget.TextView;
 
 import com.example.android.systemupdatersample.R;
@@ -49,7 +50,6 @@ public class MainActivity extends Activity {
     private static final String TAG = "MainActivity";
 
     private TextView mTextViewBuild;
-    private Spinner mSpinnerConfigs;
     private TextView mTextViewConfigsDirHint;
     private Button mButtonReload;
     private Button mButtonApplyConfig;
@@ -62,7 +62,7 @@ public class MainActivity extends Activity {
     private TextView mTextViewEngineStatus;
     private TextView mTextViewEngineErrorCode;
     private TextView mTextViewUpdateInfo;
-    private Button mButtonSwitchSlot;
+    private Button mButtonRebootDevice;
 
     private List<UpdateConfig> mConfigs;
 
@@ -75,7 +75,6 @@ public class MainActivity extends Activity {
         setContentView(R.layout.activity_main);
 
         this.mTextViewBuild = findViewById(R.id.textViewBuild);
-        this.mSpinnerConfigs = findViewById(R.id.spinnerConfigs);
         this.mTextViewConfigsDirHint = findViewById(R.id.textViewConfigsDirHint);
         this.mButtonReload = findViewById(R.id.buttonReload);
         this.mButtonApplyConfig = findViewById(R.id.buttonApplyConfig);
@@ -88,24 +87,24 @@ public class MainActivity extends Activity {
         this.mTextViewEngineStatus = findViewById(R.id.textViewEngineStatus);
         this.mTextViewEngineErrorCode = findViewById(R.id.textViewEngineErrorCode);
         this.mTextViewUpdateInfo = findViewById(R.id.textViewUpdateInfo);
-        this.mButtonSwitchSlot = findViewById(R.id.buttonSwitchSlot);
-
-        this.mTextViewConfigsDirHint.setText(UpdateConfigs.getConfigsRoot(this));
+        this.mButtonRebootDevice = findViewById(R.id.buttonRebootDevice);
 
         uiResetWidgets();
         loadUpdateConfigs();
+		if (mConfigs.size() > 0)
+			this.mTextViewConfigsDirHint.setText(mConfigs.get(0).getName());
 
         this.mUpdateManager.setOnStateChangeCallback(this::onUpdaterStateChange);
         this.mUpdateManager.setOnEngineStatusUpdateCallback(this::onEngineStatusUpdate);
         this.mUpdateManager.setOnEngineCompleteCallback(this::onEnginePayloadApplicationComplete);
         this.mUpdateManager.setOnProgressUpdateCallback(this::onProgressUpdate);
-    }
+	}
 
-    @Override
-    protected void onDestroy() {
-        this.mUpdateManager.setOnEngineStatusUpdateCallback(null);
-        this.mUpdateManager.setOnProgressUpdateCallback(null);
-        this.mUpdateManager.setOnEngineCompleteCallback(null);
+	@Override
+	protected void onDestroy() {
+		this.mUpdateManager.setOnEngineStatusUpdateCallback(null);
+		this.mUpdateManager.setOnProgressUpdateCallback(null);
+		this.mUpdateManager.setOnEngineCompleteCallback(null);
         super.onDestroy();
     }
 
@@ -128,35 +127,35 @@ public class MainActivity extends Activity {
      */
     public void onReloadClick(View view) {
         loadUpdateConfigs();
-    }
-
-    /**
-     * view config button is clicked
-     */
-    public void onViewConfigClick(View view) {
-        UpdateConfig config = mConfigs.get(mSpinnerConfigs.getSelectedItemPosition());
-        new AlertDialog.Builder(this)
-                .setTitle(config.getName())
-                .setMessage(config.getRawJson())
-                .setPositiveButton(R.string.close, (dialog, id) -> dialog.dismiss())
-                .show();
+		if (mConfigs.size() > 0)
+			this.mTextViewConfigsDirHint.setText(mConfigs.get(0).getName());
+		else
+			this.mTextViewConfigsDirHint.setText(R.string.package_info);
     }
 
     /**
      * apply config button is clicked
      */
     public void onApplyConfigClick(View view) {
-        new AlertDialog.Builder(this)
-                .setTitle("Apply Update")
-                .setMessage("Do you really want to apply this update?")
-                .setIcon(android.R.drawable.ic_dialog_alert)
-                .setPositiveButton(android.R.string.ok, (dialog, whichButton) -> {
-                    uiResetWidgets();
-                    uiResetEngineText();
-                    applyUpdate(getSelectedConfig());
-                })
-                .setNegativeButton(android.R.string.cancel, null)
-                .show();
+		if (mConfigs.size() <= 0)
+        	new AlertDialog.Builder(this)
+            	    .setTitle("安装升级包")
+                	.setMessage("未发现有效的升级包，请确定插入U盘！")
+                	.setIcon(android.R.drawable.ic_dialog_alert)
+                	.setPositiveButton(android.R.string.ok, null)
+                	.show();
+		else
+        	new AlertDialog.Builder(this)
+            	    .setTitle("安装升级包")
+                	.setMessage("请确定要安装此升级包？")
+                	.setIcon(android.R.drawable.ic_dialog_alert)
+                	.setPositiveButton(android.R.string.ok, (dialog, whichButton) -> {
+                    	uiResetWidgets();
+                    	uiResetEngineText();
+                    	applyUpdate(getSelectedConfig());
+                	})
+                	.setNegativeButton(android.R.string.cancel, null)
+                	.show();
     }
 
     private void applyUpdate(UpdateConfig config) {
@@ -172,8 +171,8 @@ public class MainActivity extends Activity {
      */
     public void onStopClick(View view) {
         new AlertDialog.Builder(this)
-                .setTitle("Stop Update")
-                .setMessage("Do you really want to cancel running update?")
+                .setTitle("停止升级")
+                .setMessage("确定取消此次升级？ 你可以点击 \'安装升级包\' 按钮继续升级。")
                 .setIcon(android.R.drawable.ic_dialog_alert)
                 .setPositiveButton(android.R.string.ok, (dialog, whichButton) -> {
                     cancelRunningUpdate();
@@ -239,9 +238,15 @@ public class MainActivity extends Activity {
     /**
      * switch slot button clicked
      */
-    public void onSwitchSlotClick(View view) {
-        uiResetWidgets();
-        mUpdateManager.setSwitchSlotOnReboot();
+    public void onRebootDeviceClick(View view) {
+        try {
+			PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
+			if (pm != null) {
+				pm.reboot(null);
+			}
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to reboot device ", e);
+        }
     }
 
     /**
@@ -300,6 +305,12 @@ public class MainActivity extends Activity {
                         + " " + completionState);
         runOnUiThread(() -> {
             setUiEngineErrorCode(errorCode);
+			if (completionState.equals("SUCCESS")) {
+				mButtonRebootDevice.setVisibility(View.VISIBLE);
+				mButtonRebootDevice.setEnabled(true);
+				mTextViewUpdateInfo.setVisibility(View.VISIBLE);
+				mTextViewUpdateInfo.setEnabled(true);
+			}
         });
     }
 
@@ -313,7 +324,6 @@ public class MainActivity extends Activity {
     /** resets ui */
     private void uiResetWidgets() {
         mTextViewBuild.setText(Build.DISPLAY);
-        mSpinnerConfigs.setEnabled(false);
         mButtonReload.setEnabled(false);
         mButtonApplyConfig.setEnabled(false);
         mButtonStop.setEnabled(false);
@@ -322,20 +332,21 @@ public class MainActivity extends Activity {
         mButtonResume.setEnabled(false);
         mProgressBar.setEnabled(false);
         mProgressBar.setVisibility(ProgressBar.INVISIBLE);
-        mButtonSwitchSlot.setEnabled(false);
+        mButtonRebootDevice.setEnabled(false);
+		mButtonRebootDevice.setVisibility(View.GONE);
         mTextViewUpdateInfo.setTextColor(Color.parseColor("#aaaaaa"));
+		mTextViewUpdateInfo.setVisibility(View.GONE);
     }
 
     private void uiResetEngineText() {
         mTextViewEngineStatus.setText(R.string.unknown);
-        mTextViewEngineErrorCode.setText(R.string.unknown);
+        mTextViewEngineErrorCode.setText(R.string.none);
         // Note: Do not reset mTextViewUpdaterState; UpdateManager notifies updater state properly.
     }
 
     private void uiStateIdle() {
         uiResetWidgets();
         mButtonReset.setEnabled(true);
-        mSpinnerConfigs.setEnabled(true);
         mButtonReload.setEnabled(true);
         mButtonApplyConfig.setEnabled(true);
         mProgressBar.setProgress(0);
@@ -362,8 +373,10 @@ public class MainActivity extends Activity {
         mButtonReset.setEnabled(true);
         mProgressBar.setEnabled(true);
         mProgressBar.setVisibility(ProgressBar.VISIBLE);
-        mButtonSwitchSlot.setEnabled(true);
+        mButtonRebootDevice.setEnabled(true);
+        mButtonRebootDevice.setVisibility(View.VISIBLE);
         mTextViewUpdateInfo.setTextColor(Color.parseColor("#777777"));
+        mTextViewUpdateInfo.setVisibility(View.VISIBLE);
     }
 
     private void uiStateError() {
@@ -383,7 +396,6 @@ public class MainActivity extends Activity {
      */
     private void loadUpdateConfigs() {
         mConfigs = UpdateConfigs.getUpdateConfigs(this);
-        loadConfigsToSpinner(mConfigs);
     }
 
     /**
@@ -410,18 +422,8 @@ public class MainActivity extends Activity {
         mTextViewUpdaterState.setText(stateText + "/" + state);
     }
 
-    private void loadConfigsToSpinner(List<UpdateConfig> configs) {
-        String[] spinnerArray = UpdateConfigs.configsToNames(configs);
-        ArrayAdapter<String> spinnerArrayAdapter = new ArrayAdapter<>(this,
-                android.R.layout.simple_spinner_item,
-                spinnerArray);
-        spinnerArrayAdapter.setDropDownViewResource(android.R.layout
-                .simple_spinner_dropdown_item);
-        mSpinnerConfigs.setAdapter(spinnerArrayAdapter);
-    }
-
     private UpdateConfig getSelectedConfig() {
-        return mConfigs.get(mSpinnerConfigs.getSelectedItemPosition());
+        return mConfigs.get(0);
     }
 
 }
diff --git a/updater_sample/src/com/example/android/systemupdatersample/util/UpdateConfigs.java b/updater_sample/src/com/example/android/systemupdatersample/util/UpdateConfigs.java
index bbefcaf1..f816de98 100644
--- a/updater_sample/src/com/example/android/systemupdatersample/util/UpdateConfigs.java
+++ b/updater_sample/src/com/example/android/systemupdatersample/util/UpdateConfigs.java
@@ -22,6 +22,9 @@ import android.util.Log;
 import com.example.android.systemupdatersample.UpdateConfig;
 
 import java.io.File;
+import java.io.FileInputStream;
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Paths;
@@ -30,12 +33,16 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Optional;
 
+
 /**
  * Utility class for working with json update configurations.
  */
 public final class UpdateConfigs {
 
     public static final String UPDATE_CONFIGS_ROOT = "configs/";
+    public static final String UPDATE_MEDIA_ROOT = "/mnt/media_rw";
+    public static final String UPDATE_INFO_FILE = "ota_info.txt";
+    public static final String ZEEKR_OTA_META = "OTA PACKAGE NAME";
 
     /**
      * @param configs update configs
@@ -60,34 +67,54 @@ public final class UpdateConfigs {
      * @return list of configs from directory {@link UpdateConfigs#getConfigsRoot}
      */
     public static List<UpdateConfig> getUpdateConfigs(Context context) {
-        File root = new File(getConfigsRoot(context));
+        File root = new File(UPDATE_MEDIA_ROOT);
         ArrayList<UpdateConfig> configs = new ArrayList<>();
+		boolean findConfig = false;
         if (!root.exists()) {
             return configs;
         }
         for (final File f : root.listFiles()) {
-            if (!f.isDirectory() && f.getName().endsWith(".json")) {
-                try {
-                    String json = new String(Files.readAllBytes(f.toPath()),
-                            StandardCharsets.UTF_8);
-                    configs.add(UpdateConfig.fromJson(json));
-                } catch (Exception e) {
-                    Log.e("UpdateConfigs", "Can't read/parse config file " + f.getName(), e);
-                    throw new RuntimeException(
-                            "Can't read/parse config file " + f.getName(), e);
-                }
-            }
-        }
-        return configs;
-    }
+			findConfig = false;
+			if (f.isDirectory()) {
+				for (final File f2: f.listFiles()) {
+					if (!f2.isDirectory() && f2.getName().equals(UPDATE_INFO_FILE)) {
+						try {
+							FileInputStream inputStream = new FileInputStream(f2);
+							BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
+							String line;
 
-    /**
-     * @param filename searches by given filename
-     * @param config searches in {@link UpdateConfig#getAbConfig()}
-     * @return offset and size of {@code filename} in the package zip file
-     *         stored as {@link UpdateConfig.PackageFile}.
-     */
-    public static Optional<UpdateConfig.PackageFile> getPropertyFile(
+							if ((line = reader.readLine()) != null && line.equals(ZEEKR_OTA_META)) {
+								if ((line = reader.readLine()) != null) {
+									File pack = new File(UPDATE_MEDIA_ROOT+"/"+ f.getName() + "/" + line);
+									if (pack.exists()) {
+										Log.d("### updater", pack.getName() + " exists");
+										String packName = f.getName() + "/" + line;
+										String packUrl = "file://" + UPDATE_MEDIA_ROOT + "/" + packName;
+										configs.add(new UpdateConfig(packName, packUrl, UpdateConfig.AB_INSTALL_TYPE_NON_STREAMING));
+									}
+									findConfig = true;
+								}
+							}
+							reader.close();
+						} catch (Exception e) {
+							e.printStackTrace();
+						}
+						if (findConfig)
+							break;
+					}
+				}
+			}
+		}
+		return configs;
+	}
+
+	/**
+	 * @param filename searches by given filename
+	 * @param config searches in {@link UpdateConfig#getAbConfig()}
+	 * @return offset and size of {@code filename} in the package zip file
+	 *         stored as {@link UpdateConfig.PackageFile}.
+	 */
+	public static Optional<UpdateConfig.PackageFile> getPropertyFile(
             final String filename,
             UpdateConfig config) {
         return Arrays
-- 
2.25.1

